### PR TITLE
Fix CLogicalGbAgg invalid xform candidate

### PR DESCRIPTION
**Pull request by dgkimura**
CXformSplitGbAgg xform splits a "global" aggregate operator into a pair
of local and global aggregates. However, it doesn't make sense to split
a "local" aggregate and doing so can lead to unexpected states.

For example, the following SQL:
    ```sql
    CREATE TABLE a_table(col text);

    SELECT a.col FROM (
        SELECT col FROM a_table e
        WHERE exists
        (
            SELECT col FROM a_table
        )
    ) a
    WHERE exists (
        SELECT c.col FROM a_table c
        INNER JOIN a_table b
        ON c.col = a.col
        WHERE exists (
            SELECT distinct _a.col FROM (
                SELECT distinct col FROM a_table _a
            ) _a
            WHERE _a.col = c.col
        )
    );
    ```

Leads to incorrectly using CXformSimplifyGbAgg on a "local" aggregate
which is unexpected, as asserted in CXformSimplifyGbAgg::Exfp().

--

Note: I rolled in an unrelated unittest for commit 3da1cc2

### CODE REVIEW

File: src/backend/gporca/server/src/unittest/gpopt/xforms/CXformRightOuterJoin2HashJoinTest.cpp [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+GPOS_RESULT
+CXformRightOuterJoin2HashJoinTest::EresUnittest_Transform()
+{
+	CAutoMemoryPool amp;
+	CMemoryPool *mp = amp.Pmp();
+
+	CMDProviderMemory *pmdp = CTestUtils::m_pmdpf;
+	pmdp->AddRef();
+	CMDAccessor mda(mp, CMDCache::Pcache(), CTestUtils::m_sysidDefault, pmdp);
+	CAutoOptCtxt aoc(mp, &mda, NULL, /* pceeval */
```
Comment (d) [inline src/backend/gporca/server/src/unittest/gpopt/xforms/CXformRightOuterJoin2HashJoinTest.cpp:59]:
```suggestion
	CAutoOptCtxt aoc(mp, &mda, nullptr, /* pceeval */
```

File: src/backend/gporca/server/src/unittest/gpopt/xforms/CXformRightOuterJoin2HashJoinTest.cpp [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+	CXformResult *result = GPOS_NEW(mp) CXformResult(mp);
+
+	CXformRightOuterJoin2HashJoin *xform =
+		GPOS_NEW(mp) CXformRightOuterJoin2HashJoin(mp);
+
+	CExpression *pLogicalRightOuterJoin = CTestUtils::PexprRightOuterJoin(mp);
+
+	// xform should work even if child stats aren't derived yet.
+	GPOS_ASSERT(NULL == (*pLogicalRightOuterJoin)[0]->Pstats());
+	GPOS_ASSERT(NULL == (*pLogicalRightOuterJoin)[1]->Pstats());
```
Comment (d) [inline src/backend/gporca/server/src/unittest/gpopt/xforms/CXformRightOuterJoin2HashJoinTest.cpp:72]:
```suggestion
	GPOS_ASSERT(nullptr == (*pLogicalRightOuterJoin)[0]->Pstats());
	GPOS_ASSERT(nullptr == (*pLogicalRightOuterJoin)[1]->Pstats());
```

File: src/backend/gporca/server/include/unittest/gpopt/operators/CLogicalGbAggTest.h [hunk 3]
```diff
@@ -0,0 +1,40 @@
+//---------------------------------------------------------------------------
+//	Greenplum Database
+//	Copyright (C) 2012 EMC Corp.
```
Comment (bhuvnesh2703) [inline src/backend/gporca/server/include/unittest/gpopt/operators/CLogicalGbAggTest.h:3]: are we still using EmC copyright?

File: src/backend/gporca/server/src/unittest/gpopt/operators/CLogicalGbAggTest.cpp [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+//---------------------------------------------------------------------------
+//	@function:
+//		CLogicalGbAggTest::EresUnittest_PxfsCandidates
+//
+//	@doc:
+//		Test set of generated xform candidates
+//
+//---------------------------------------------------------------------------
+GPOS_RESULT
+CLogicalGbAggTest::EresUnittest_PxfsCandidates()
```
Comment (bhuvnesh2703) [inline src/backend/gporca/server/src/unittest/gpopt/operators/CLogicalGbAggTest.cpp:41]: do we also want to test for EgbaggtypeIntermediate?

File: src/backend/gporca/server/include/unittest/gpopt/CTestUtils.h [hunk 5]
```diff
@@ -352,6 +352,9 @@ class CTestUtils
 	// generate left outer join on top of n-ary join expression
 	static CExpression *PexprLeftOuterJoinOnNAryJoin(CMemoryPool *mp);
 
+	// generate right outer join expression
+	static CExpression *PexprRightOuterJoin(CMemoryPool *mp);
```
Comment (bhuvnesh2703) [inline src/backend/gporca/server/include/unittest/gpopt/CTestUtils.h:356]: does this pR need right outer join stuff?

**Review by bhuvnesh2703 — APPROVED** [review]
LGTM, but there are minor comments which should be fixed before merging.

File: src/backend/gporca/server/src/unittest/gpopt/xforms/CXformRightOuterJoin2HashJoinTest.cpp (hunk 1 above)
Comment (dgkimura) [reply src/backend/gporca/server/src/unittest/gpopt/xforms/CXformRightOuterJoin2HashJoinTest.cpp:59]: Thanks Jesse.  Super elated to see that you still watch the repo.

File: src/backend/gporca/server/include/unittest/gpopt/operators/CLogicalGbAggTest.h (hunk 3 above)
Comment (dgkimura) [reply src/backend/gporca/server/include/unittest/gpopt/operators/CLogicalGbAggTest.h]: Yeah, seems so.
```
$ git grep "Copyright .*EMC" | wc -l
1174
$ git grep "Copyright .*EMC" src/backend/gporca/ | wc -l
1095
```

I updated references in this PR.

File: src/backend/gporca/server/src/unittest/gpopt/operators/CLogicalGbAggTest.cpp (hunk 4 above)
Comment (dgkimura) [reply src/backend/gporca/server/src/unittest/gpopt/operators/CLogicalGbAggTest.cpp:41]: Sure, added it.

File: src/backend/gporca/server/include/unittest/gpopt/CTestUtils.h (hunk 5 above)
Comment (dgkimura) [reply src/backend/gporca/server/include/unittest/gpopt/CTestUtils.h:356]: It's not needed.  I was piggybacking https://github.com/greenplum-db/gpdb/pull/12999/commits/92e80846fd837a14ebbe5b3005415c3d285d072f on this PR.

**Pull request merged**